packages/canvas/canvas-graphics/src/CanvasGraphicsRenderer.ts

```diff
@@ -73,6 +73,118 @@ export class CanvasGraphicsRenderer
+     * @private
+     * @param {PIXI.FillStyle} fillStyle
+     * @param {PIXI.LineStyle} lineStyle
+     * @param {string|CanvasPattern} contextFillStyle
+     * @param {string|CanvasPattern} contextStrokeStyle
+     * @param {number} worldAlpha
+     */
+    private draw(
+        fillStyle: FillStyle,
+        lineStyle: LineStyle,
+        contextFillStyle: string|CanvasPattern,
+        contextStrokeStyle: string|CanvasPattern,
+        worldAlpha: number
+    ): void
+    {
+        const context = this.renderer.context;
+        const innerStrokeAlignment = lineStyle.visible && lineStyle.alignment === 0;
+        const outerStrokeAlignment = lineStyle.visible && lineStyle.alignment === 1;
+
+        if (fillStyle.visible && !outerStrokeAlignment)
+        {
+            context.globalAlpha = fillStyle.alpha * worldAlpha;
+            context.fillStyle = contextFillStyle;
+            context.fill();
+        }
+
+        if (lineStyle.visible)
+        {
+            if (innerStrokeAlignment)
+            {
+                context.save();
+                context.clip();
+            }
+
+            context.globalAlpha = lineStyle.alpha * worldAlpha;
+            context.strokeStyle = contextStrokeStyle;
+            context.stroke();
+
+            if (innerStrokeAlignment)
+            {
+                context.restore();
+            }
+
+            if (outerStrokeAlignment)
+            {
+                context.globalAlpha = fillStyle.alpha * worldAlpha;
+                context.fillStyle = contextFillStyle;
+                context.fill();
+            }
+        }
+    }
+
+    /**
+     * draw rectangle on canvas
+     *
+     * @private
+     * @param {PIXI.Rectangle} rectangle
+     * @param {PIXI.FillStyle} fillStyle
+     * @param {PIXI.LineStyle} lineStyle
+     * @param {string|CanvasPattern} contextFillStyle
+     * @param {string|CanvasPattern} contextStrokeStyle
+     * @param {number} worldAlpha
+     */
+    private drawRect(
+        rectangle: Rectangle,
+        fillStyle: FillStyle,
+        lineStyle: LineStyle,
+        contextFillStyle: string|CanvasPattern,
+        contextStrokeStyle: string|CanvasPattern,
+        worldAlpha: number
+    ): void
+    {
+        const context = this.renderer.context;
+        const innerStrokeAlignment = lineStyle.visible && lineStyle.alignment === 0;
+        const outerStrokeAlignment = lineStyle.visible && lineStyle.alignment === 1;
+
+        if (fillStyle.visible && !outerStrokeAlignment)
+        {
+            context.globalAlpha = fillStyle.alpha * worldAlpha;
+            context.fillStyle = contextFillStyle;
+            context.fillRect(rectangle.x, rectangle.y, rectangle.width, rectangle.height);
+        }
+
+        if (lineStyle.visible)
+        {
+            if (innerStrokeAlignment)
+            {
+                context.save();
+                context.rect(rectangle.x, rectangle.y, rectangle.width, rectangle.height);
+                context.clip();
```

Comment:
Instead of using clipping for rectangles, how about we draw two rectangles and the `evenodd` fill rule to do a stroke? That would be faster since the Canvas2D implementation won't have to do clipping calculations.